src/Bicep.Core/TypeSystem/ResourceTypeProviderFactory.cs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using Azure.Bicep.Types;
+using Azure.Bicep.Types.Az;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Features;
+using Bicep.Core.Modules;
+using Bicep.Core.Registry.Oci;
+using Bicep.Core.Semantics.Namespaces;
+using Bicep.Core.TypeSystem.Az;
+using Microsoft.WindowsAzure.ResourceStack.Common.EventSources.ManifestInstaller;
+using Newtonsoft.Json;
+
+namespace Bicep.Core.TypeSystem
+{
+    public class ResourceTypeProviderFactory : IResourceTypeProviderFactory
+    {
+        private record ResourceTypeLoaderKey(string Name, string Version);
+        private readonly ResourceTypeLoaderKey BuiltInAzResourceTypeLoaderKey = new(AzNamespaceType.BuiltInName, AzNamespaceType.Settings.ArmTemplateProviderVersion);
+        private readonly Dictionary<ResourceTypeLoaderKey, IResourceTypeProvider> cachedResourceTypeLoaders;
+
+        public ResourceTypeProviderFactory()
+        {
+            cachedResourceTypeLoaders = new() {
+                {BuiltInAzResourceTypeLoaderKey, new AzResourceTypeProvider(new AzResourceTypeLoader(new AzTypeLoader()),AzNamespaceType.Settings.ArmTemplateProviderVersion)},
+            };
+        }
+
+        public ResultWithDiagnostic<IResourceTypeProvider> GetResourceTypeProvider(ResourceTypesProviderDescriptor providerDescriptor, IFeatureProvider features)
+        {
+            if (!features.DynamicTypeLoadingEnabled)
+            {
+                return new(cachedResourceTypeLoaders[BuiltInAzResourceTypeLoaderKey]);
+            }
+            var key = new ResourceTypeLoaderKey(providerDescriptor.Alias, providerDescriptor.Version);
+
+            if (cachedResourceTypeLoaders.ContainsKey(key))
+            {
+                return new(cachedResourceTypeLoaders[key]);
+            }
+            // should neverbe null since provider restore success is validated prior.
+            string providerDirectory = Path.GetDirectoryName(providerDescriptor.Path) ?? throw new UnreachableException("the provider directory doesn't exist");
```

Comment:
We should avoid using types from `System.IO` directly within `Bicep.Core`. Replace the usage of `Path`, `File`, etc. with `IFileResolver` or `IFileSystem`.

Reply:
To be addressed as a separate PR